packages/wdio-appium-service/src/launcher.js

```diff
@@ -0,0 +1,117 @@
+
+const log = logger('@wdio/appium-service')
+const DEFAULT_LOG_FILENAME = 'appium.txt'
+
+export class AppiumLauncher {
+    constructor() {
+        this.logPath = null
+        this.command = ''
+        this.appiumArgs = []
+        this.waitStartTime = 0
+    }
+
+    async onPrepare(config) {
+        const appiumConfig = config.appium || {}
+
+        this.logPath = appiumConfig.logPath
+        this.command = appiumConfig.command || this._getAppiumCommand()
+        this.appiumArgs = this._cliArgsFromKeyValue(appiumConfig.args || {})
+        this.waitStartTime = appiumConfig.waitStartTime || 1000
+
+        const asyncStartAppium = promisify(this._startAppium)
+        this.process = await asyncStartAppium(this.command, this.appiumArgs, this.waitStartTime)
+
+        if (typeof this.logPath === 'string') {
+            this._redirectLogStream(this.logPath)
+        }
+    }
+
+    onComplete() {
+        if(this.process) {
+            log.debug(`Appium (pid: ${process.pid}) killed`)
+            this.process.kill()
+        }
+    }
+
+    _startAppium(command, args, waitStartTime, callback) {
+        log.debug(`Will spawn Appium process: ${command} ${args.join(' ')}`)
+        let process = spawn(command, args, { stdio: ['ignore', 'pipe', 'pipe'] })
+
+        const exitCallback = (exitCode) => {
+            clearTimeout(timer)
+            callback(new Error(`Appium exited before timeout (exit code: ${exitCode})`), null)
+        }
+
+        const timer = setTimeout(() => {
+            process.removeListener('exit', exitCallback)
+            log.debug(`Appium started with ID: ${process.pid}`)
+            callback(null, process)
+        }, waitStartTime)
+
+        process.once('exit', exitCallback)
+    }
+
+    _redirectLogStream(logPath) {
+        const logFile = getFilePath(logPath, DEFAULT_LOG_FILENAME)
+
+        // ensure file & directory exists
+        ensureFileSync(logFile)
+
+        log.debug(`Appium logs written to: ${logFile}`)
+        const logStream = createWriteStream(logFile, { flags: 'w' })
+        this.process.stdout.pipe(logStream)
+        this.process.stderr.pipe(logStream)
+    }
+
+    _getAppiumCommand() {
+        return join('node_modules', '.bin', 'appium')
+    }
+
+    _cliArgsFromKeyValue(keyValueArgs) {
+        const cliArgs = []
+        for (let key in keyValueArgs) {
+            const value = keyValueArgs[key]
+            // If the value is false or null the argument is discarded
+            if ((typeof value === 'boolean' && !value) || value === null) {
+                continue
+            }
+
+            cliArgs.push(this._lowerCamelToCliOptionName(key))
+
+            // Only non-boolean and non-null values are added as option values
+            if (typeof value !== 'boolean' && value !== null) {
+                cliArgs.push(this._sanitizeCliOptionValue(value))
+            }
+        }
+        return cliArgs
+    }
+
+    _lowerCamelToCliOptionName(camelCasedKey) {
```

Comment:
Can we replace this with https://github.com/blakeembrey/camel-case

Reply:
I used `param-case` as it needs dashes instead of spaces.